File: android/app/src/main/java/com/reaxn/FaceMeshFrameProcessor.kt
package com.reaxn

import com.google.mediapipe.framework.image.MediaImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.core.Delegate
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker.FaceLandmarkerOptions
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy

// VisionCamera frame-processor plugin: runs MediaPipe Face Landmarker (478-pt mesh) per frame and
// reduces it — natively, to keep the JS bridge cheap — to the SAME 6-anchor contract the lighter
// BlazeFace plugin returns, so the JS orientation/reduce path is shared and unchanged:
//   { points: [[x,y]×6] }  →  [0]=right eye [1]=left eye [2]=nose [3]=mouth [4]=right cheek [5]=left cheek
// Plus, on request (mesh lenses), the full 478-pt mesh: { mesh: [[x,y]×478] }. Blendshapes are NOT
// emitted — they ran on a CPU model; JS now derives the few expression signals it uses from the mesh.
// Registered under the JS name "faceMesh" (MainApplication). Needs assets/face_landmarker.task.
// Orientation is handled in JS (faceTracking.ts) — keypoints are returned in the raw sensor space.
class FaceMeshFrameProcessor(proxy: VisionCameraProxy, options: Map<String, Any>?) :
    FrameProcessorPlugin() {

  init { warmUp(proxy.context) }

  override fun callback(frame: Frame, arguments: Map<String, Any>?): Any? {
    val landmarker = detector ?: return mapOf("err" to "no_model")
    val image = frame.image ?: return mapOf("err" to "no_image")
    val mp = MediaImageBuilder(image).build()
    // IMAGE mode: per-frame detect (no timestamp). VIDEO mode: tracked detect with a monotonic ts.
    val result = try {
      if (USE_IMAGE_MODE) landmarker.detect(mp)
      else landmarker.detectForVideo(mp, android.os.SystemClock.uptimeMillis())
    } catch (e: Throwable) { return mapOf("err" to "detect_fail") }
    val faces = result.faceLandmarks()
    if (faces.isEmpty() || faces[0].size < 468) return mapOf("err" to "no_face")
    val f = faces[0]

    // Canonical MediaPipe FaceMesh indices → 6 BlazeFace-equivalent anchors (subject-anatomical, image
    // coords). Eye/mouth anchors average two contour corners for a stable center.
    fun pt(i: Int) = listOf(f[i].x().toDouble(), f[i].y().toDouble())
    fun mid(a: Int, b: Int) = listOf(((f[a].x() + f[b].x()) / 2f).toDouble(), ((f[a].y() + f[b].y()) / 2f).toDouble())
    val points = listOf(
      mid(33, 133),   // right eye
      mid(263, 362),  // left eye
      pt(1),          // nose tip
      mid(13, 14),    // mouth
      pt(234),        // right cheek
      pt(454),        // left cheek
    )
    val out = HashMap<String, Any>()
    out["points"] = points

    // Full 478-pt mesh, only when JS asks (mesh lenses) — keeps the bridge cheap by default. FLAT
    // [x0,y0,x1,y1,…]: one numeric array marshals far cheaper than 478 nested arrays. reduce() de-interleaves.
    if (arguments?.get("mesh") == true) {
      val mesh = ArrayList<Double>(f.size * 2)
      for (p in f) { mesh.add(p.x().toDouble()); mesh.add(p.y().toDouble()) }
      out["mesh"] = mesh
    }

    return out
  }

  companion object {
    // DETECTOR MODE. VIDEO mode tracks the face across frames (a per-frame detection shortcut) — far
    // lighter, which matters on Android GPUs/CPUs. IMAGE mode runs a FULL detection every frame: heavier,
    // and on slower Android devices it falls badly behind (the mesh lags the face). We default to VIDEO
    // now that the JS back-pressure gate handles the runOnJS backlog that IMAGE was briefly used to probe;
    // VIDEO's mild temporal smoothing is also steadier when idle. Set true only to A/B. Rebuild to apply.
    const val USE_IMAGE_MODE = false

    // One shared landmarker so the launch-time warm-up and the frame processor load the model once.
    @Volatile
    var detector: FaceLandmarker? = null
      private set

    @JvmStatic
    fun warmUp(context: android.content.Context) {
      if (detector != null) { return }
      synchronized(this) {
        if (detector != null) { return }
        detector = build(context, Delegate.GPU)?.also { android.util.Log.i("FaceMesh", "delegate=GPU") }
          ?: build(context, Delegate.CPU)?.also { android.util.Log.i("FaceMesh", "delegate=CPU") }
          ?: run { android.util.Log.w("FaceMesh", "delegate=none (model failed to load)"); null }
      }
    }

    private fun build(context: android.content.Context, delegate: Delegate): FaceLandmarker? {
      return try {
        val base = BaseOptions.builder()
          .setModelAssetPath("face_landmarker.task")
          .setDelegate(delegate)
          .build()
        val opts = FaceLandmarkerOptions.builder()
          .setBaseOptions(base)
          .setRunningMode(if (USE_IMAGE_MODE) RunningMode.IMAGE else RunningMode.VIDEO)
          .setNumFaces(1)
          .setMinFaceDetectionConfidence(0.5f)
          .setMinFacePresenceConfidence(0.5f)
          .setMinTrackingConfidence(0.5f)
          // Blendshapes OFF: MediaPipe runs them on a CPU (XNNPACK) model every frame regardless of
          // delegate, and we only used a few signals — now derived from the GPU mesh geometry in
          // faceFrame (smile/brow) or the anchor proxy (mouthOpen). Trims per-frame inference everywhere.
          .setOutputFaceBlendshapes(false)
          // The 4×4 facial-transform matrix (a per-frame PnP solve) is not consumed in JS — skip it to
          // save inference time. Re-enable if a lens ever needs head pose.
          .setOutputFacialTransformationMatrixes(false)
          .build()
        FaceLandmarker.createFromOptions(context.applicationContext, opts)
      } catch (e: Throwable) { null }
    }
  }
}

File: ios/FaceMeshFrameProcessor.swift
import VisionCamera
import MediaPipeTasksVision
import Foundation
import QuartzCore

// VisionCamera frame-processor plugin: runs MediaPipe **Face Landmarker** (478-pt mesh) per frame and
// reduces it — on the native side, to keep the JS bridge cheap — to the SAME 6-anchor contract the
// lighter BlazeFace plugin returns, so the JS orientation/reduce path is shared and unchanged:
//   { points: [[x,y]×6] }  →  [0]=right eye [1]=left eye [2]=nose [3]=mouth [4]=right cheek [5]=left cheek
// Plus, on request (mesh lenses), the full 478-pt mesh: { mesh: [[x,y]×478] }. Blendshapes are NOT
// emitted — they ran on a CPU model; JS now derives the few expression signals it uses from the mesh.
// Registered under the JS name "faceMesh" (see FaceMeshFrameProcessor.m). Needs `face_landmarker.task`
// bundled (Build Phases → Copy Bundle Resources). VIDEO mode tracks across frames with a monotonic ts.
@objc public final class FaceMeshShared: NSObject {
  @objc public static let shared = FaceMeshShared()
  private let lock = NSLock()
  public private(set) var landmarker: FaceLandmarker?

  // SNAPPINESS TOGGLE. IMAGE mode detects each frame independently — instant 1:1 tracking, no temporal
  // smoothing (the mesh never trails / "slides" to catch up on a fast move), at the cost of a little
  // more idle jitter and heavier per-frame compute (full detection every frame, no tracking shortcut).
  // VIDEO mode tracks + stabilizes across frames: rock-steady when idle, but lags fast motion. Flip to
  // false to restore VIDEO. Must rebuild the app to take effect.
  @objc public static let useImageMode = true

  @objc public func warmUp() {
    lock.lock(); defer { lock.unlock() }
    if landmarker != nil { return }
    guard let path = Bundle.main.path(forResource: "face_landmarker", ofType: "task") else {
      NSLog("[FaceMesh] model not found in bundle")
      return
    }
    if let gpu = Self.build(path: path, delegate: .GPU) {
      landmarker = gpu
      NSLog("[FaceMesh] delegate=GPU")
    } else {
      landmarker = Self.build(path: path, delegate: .CPU)
      NSLog("[FaceMesh] delegate=\(landmarker != nil ? "CPU" : "none")")
    }
  }

  private static func build(path: String, delegate: Delegate) -> FaceLandmarker? {
    let opts = FaceLandmarkerOptions()
    opts.baseOptions.modelAssetPath = path
    opts.baseOptions.delegate = delegate
    opts.runningMode = useImageMode ? .image : .video
    opts.numFaces = 1
    opts.minFaceDetectionConfidence = 0.5
    opts.minFacePresenceConfidence = 0.5
    opts.minTrackingConfidence = 0.5
    // Blendshapes OFF: MediaPipe runs them on a CPU (XNNPACK) model every frame regardless of delegate,
    // and we only used a few signals — now derived from the GPU mesh geometry in faceFrame (smile/brow)
    // or the anchor proxy (mouthOpen). Dropping the CPU model trims per-frame inference on every device.
    opts.outputFaceBlendshapes = false
    // The 4×4 facial-transform matrix (a per-frame PnP solve) is not consumed in JS — skip it to save
    // inference time. Re-enable if a lens ever needs head pose.
    opts.outputFacialTransformationMatrixes = false
    return try? FaceLandmarker(options: opts)
  }
}

@objc(FaceMeshFrameProcessor)
public class FaceMeshFrameProcessor: FrameProcessorPlugin {

  // Pinned like the BlazeFace plugin: the app is portrait-locked, so the first frame's orientation is
  // the one we feed MediaPipe for the whole session (physically rotating the phone won't flip it).
  private var lockedOrientation: UIImage.Orientation?

  public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
    super.init(proxy: proxy, options: options)
    FaceMeshShared.shared.warmUp()
  }

  // CALIBRATION: let JS override the orientation we feed MediaPipe (arg "ori") so we can dial in the
  // value that makes the mesh detect upright WITHOUT a native rebuild per attempt. Returns nil for an
  // unknown/absent string → falls back to the pinned frame orientation.
  private func uiOrientation(_ s: String?) -> UIImage.Orientation? {
    switch s {
    case "up": return .up
    case "down": return .down
    case "left": return .left
    case "right": return .right
    case "upMirrored": return .upMirrored
    case "downMirrored": return .downMirrored
    case "leftMirrored": return .leftMirrored
    case "rightMirrored": return .rightMirrored
    default: return nil
    }
  }

  public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
    guard let landmarker = FaceMeshShared.shared.landmarker else { return ["err": "no_model"] }
    if lockedOrientation == nil { lockedOrientation = frame.orientation }
    let orientation = uiOrientation(arguments?["ori"] as? String) ?? lockedOrientation ?? frame.orientation
    let image: MPImage
    do { image = try MPImage(sampleBuffer: frame.buffer, orientation: orientation) }
    catch { return ["err": "no_image"] }
    // IMAGE mode: per-frame detect (no timestamp). VIDEO mode: tracked detect with a monotonic ts.
    let detected: FaceLandmarkerResult?
    if FaceMeshShared.useImageMode {
      detected = try? landmarker.detect(image: image)
    } else {
      detected = try? landmarker.detect(videoFrame: image, timestampInMilliseconds: Int(CACurrentMediaTime() * 1000))
    }
    guard let result = detected else { return ["err": "detect_fail"] }
    guard let face = result.faceLandmarks.first, face.count >= 468 else { return ["err": "no_face"] }

    // Canonical MediaPipe FaceMesh indices → 6 BlazeFace-equivalent anchors (subject-anatomical, image
    // coords). Eye/mouth anchors average two contour corners for a stable center.
    func mid(_ a: Int, _ b: Int) -> [Double] { [Double(face[a].x + face[b].x) / 2.0, Double(face[a].y + face[b].y) / 2.0] }
    func pt(_ i: Int) -> [Double] { [Double(face[i].x), Double(face[i].y)] }
    let points: [[Double]] = [
      mid(33, 133),   // right eye  (outer 33 + inner 133)
      mid(263, 362),  // left eye   (outer 263 + inner 362)
      pt(1),          // nose tip
      mid(13, 14),    // mouth      (upper-inner 13 + lower-inner 14)
      pt(234),        // right cheek
      pt(454),        // left cheek
    ]
    var out: [String: Any] = ["points": points]

    // Full 478-pt mesh, only when JS asks (Debug lens) — keeps the bridge cheap by default.
    if let wantMesh = arguments?["mesh"] as? Bool, wantMesh {
      var mesh: [[Double]] = []; mesh.reserveCapacity(face.count)
      for p in face { mesh.append([Double(p.x), Double(p.y)]) }
      out["mesh"] = mesh
    }

    return out
  }
}
